src/Psalm/Internal/Analyzer/Statements/Expression/Call/StaticMethod/ExistingAtomicStaticCallAnalyzer.php

```diff
@@ -551,6 +551,18 @@ private static function getMethodReturnType(
             ) {
                 $static_type = $context->self;
                 $context_final = $codebase->classlike_storage_provider->get($context->self)->final;
+            } elseif ($context->calling_method_id !== null) {
+                $self_method_return = $codebase->methods->getMethodReturnType(
+                    MethodIdentifier::fromMethodIdReference($context->calling_method_id),
+                    $context->self
+                );
+                // differentiate between these cases:
+                //   1. "static" in return type comes from return type of the
+                //   method CALLING the currently analyzed static method - use $context->self.
+                //   2. "static" comes from the CALLED static method - use $fq_class_name.
+                $static_type = $self_method_return !== null && self::hasStaticInType($self_method_return)
+                    ? $context->self
+                    : $fq_class_name;
```

Comment:
What is `$self_method_return` here? It's not completely clear to me what the condition is to decide if `static` should mean `$context->self` vs `$fq_class_name`.

Reply:
Hmm, I already see where this logic is going to break. I think I have a better idea, which I'll implement a bit later.

Reply:
Also, both calling and called method could have their own `static`s, but I don't see any way to resolve it for now.

Reply:
Yeah, it would be good to make sure [this](https://psalm.dev/r/2cc402de71) works. It looks like `TypeExpander` allows `$static_class_type` to be a `TNamedObject`, so you could try turning the `"static"` string into a `TNamedObject` with `$is_static = true` somewhere and play around with it. Not sure exactly how to get that to work correctly though.

Reply:
@AndrolGenhald let's extract handling two different `static`s into a different issue.

Reply:
Opened #8257 for this.